### ISSUE TITLE
Communicate read only state from api to client for groups

### ISSUE THREAD

**Issue by hmstepanek** [hypothesis/product-backlog#580]
In the API, communicate to the client whether the user can post to a given group. Note that we cannot just use group.type === "restricted" because that is a property of the group, whereas ability to post depends on the user's permissions within the group.
This is related to https://github.com/hypothesis/product-backlog/issues/575.

**Comment by hmstepanek** [comment]
So here's what I'm thinking...We can add in to the api response (possibly in h/resources in the GroupResource) a property called readonly. This property will be computed based on the user.
```
class GroupResource(object):
    def __init__(self, group, request):
        self.request = request
        self.group = group
        self.links_service = self.request.find_service(name='group_links')

    @property
    def readonly(self):
        if ((self.request.user and self.group.type == 'restricted' 
              and  self.request.user in self.group.members)
            or (self.request.user and self.group.type == 'private' 
                  and self.request.user in self.group.members)
            or self.group.type == 'open'):
            return False
        return True

```
Then in the get request, request expand=['readonly']. This would mean that now each group in the list has a readonly property within the context of the user who is logged in.

**Comment by robertknight** [comment]
> So here's what I'm thinking...We can add in to the api response (possibly in h/resources in the GroupResource) a property called readonly. This property will be computed based on the user.

This looks like a reasonable approach to me. This does raise an issue about what the client presents for unauthenticated users. Currently it shows the "Annotations" button and clicking it creates a new annotation but tells the user that they will need to login to start editing it.

I think the only logical behaviour for the API is to indicate that a group is read-only if the user is unauthenticated. The client can use this together with the group's type to show appropriate UI.

> Then in the get request, request expand=['readonly']. 

I think there is a misunderstanding over what `expand` is for. The expand param is a way to ask for the service to return the full contents of a related resource, instead of just the ID, in the response. In this context, `readonly` is a boolean field rather than the ID of a resource, so there is nothing to expand.

**Comment by seanh** [comment]
I think we need to hold off briefly on writing any code for this.

Authorization in h is handled in a centralized way in a different part of the code, and I'm not sure we want to start making authorization decisions in these `*Resource` classes. See [the security chapter in the Pyramid docs](https://docs.pylonsproject.org/projects/pyramid//en/latest/narr/security.html), [the `h.auth` package](https://github.com/hypothesis/h/tree/master/h/auth), and also [`h/authz.py`](https://github.com/hypothesis/h/blob/master/h/authz.py). For example [we add `group:<PUBID>` Pyramid "principals" to the request](https://github.com/hypothesis/h/blob/98c6a2f89b9aedaa51639fccf0fbc4a434e8eb53/h/auth/util.py#L81-L82) for each group that the user is a member of. But this doesn't support separate read and write permissions for each group.

@hmstepanek Maybe you can spend some time looking into the above?

But I would say that any in-progress pull requests, or anything to do with getting <https://github.com/hypothesis/product-backlog/issues/572> completely finished, should be done before adding <https://github.com/hypothesis/product-backlog/issues/575>-related stuff to our work in progress.

**Comment by robertknight** [comment]
> Authorization in h is handled in a centralized way in a different part of the code, and I'm not sure we want to start making authorization decisions in these *Resource classes. 

Yes, that's a good point. I completely forgot about that this morning. One other note I'll add to seanh's comments is that `h.models.Group.__acl__` already defines which principals can perform which actions on a group.
